### PR TITLE
Set up headlamp case

### PR DESCRIPTION
**Pull request by ansjsia**
Implemented the headlamp radiation case (Monte Carlo radiation model) from the Fluent tutorial in PyAnsys.

### CODE REVIEW

**Comment by ansjsia** [comment]
@dnwillia @seanpearsonuk Hi, would you be able to advise on fixing a build issue? The documentation build is failing on line 233 of `radiation_headlamp.py` where it calls `meshing.tui.mesh.check_mesh()` to check the mesh:

```
/home/ansys/actions-runner/_work/pyfluent/pyfluent/examples/00-fluent/radiation_headlamp.py failed leaving traceback:
Traceback (most recent call last):
  File "/home/ansys/actions-runner/_work/pyfluent/pyfluent/examples/00-fluent/radiation_headlamp.py", line 233, in <module>
    meshing.tui.mesh.check_mesh()
  File "/home/ansys/actions-runner/_work/_tool/Python/3.10.8/x64/lib/python3.10/site-packages/ansys/fluent/core/services/datamodel_tui.py", line 288, in __getattribute__
    if PyMenu(self.service, path).get_child_names():
  File "/home/ansys/actions-runner/_work/_tool/Python/3.10.8/x64/lib/python3.10/site-packages/ansys/fluent/core/services/datamodel_tui.py", line 166, in get_child_names
    response = self._service.get_attribute_value(request)
  File "/home/ansys/actions-runner/_work/_tool/Python/3.10.8/x64/lib/python3.10/site-packages/ansys/fluent/core/services/error_handler.py", line 15, in func
    raise RuntimeError(ex.details()) from None
RuntimeError: menu not found
```

I can't reproduce this on my local machine, which is running version `0.18.dev0` or `ansys-fluent-core`.

**Comment by mkundu1** [comment]
> @dnwillia @seanpearsonuk Hi, would you be able to advise on fixing a build issue? The documentation build is failing on line 233 of `radiation_headlamp.py` where it calls `meshing.tui.mesh.check_mesh()` to check the mesh:
> 
> ```
> /home/ansys/actions-runner/_work/pyfluent/pyfluent/examples/00-fluent/radiation_headlamp.py failed leaving traceback:
> Traceback (most recent call last):
>   File "/home/ansys/actions-runner/_work/pyfluent/pyfluent/examples/00-fluent/radiation_headlamp.py", line 233, in <module>
>     meshing.tui.mesh.check_mesh()
>   File "/home/ansys/actions-runner/_work/_tool/Python/3.10.8/x64/lib/python3.10/site-packages/ansys/fluent/core/services/datamodel_tui.py", line 288, in __getattribute__
>     if PyMenu(self.service, path).get_child_names():
>   File "/home/ansys/actions-runner/_work/_tool/Python/3.10.8/x64/lib/python3.10/site-packages/ansys/fluent/core/services/datamodel_tui.py", line 166, in get_child_names
>     response = self._service.get_attribute_value(request)
>   File "/home/ansys/actions-runner/_work/_tool/Python/3.10.8/x64/lib/python3.10/site-packages/ansys/fluent/core/services/error_handler.py", line 15, in func
>     raise RuntimeError(ex.details()) from None
> RuntimeError: menu not found
> ```
> 
> I can't reproduce this on my local machine, which is running version `0.18.dev0` of `ansys-fluent-core`.

Can you import `headlamp_pmdb_file` instead of `headlamp_spaceclaim_file`? I'm getting the following error in a local run with docker:
```
Error: The *.scdoc file format is not supported on Linux and its equivalent /mnt/pyfluent/headlamp.scdoc.pmdb is not available. As a workaround, you can import the *.scdoc file on a Windows machine (using the 'Save PMDB' option in the Import Options dialog) and then use the generated .pmdb file as the import file on your Linux system.
```
Fluent expects `headlamp.scdoc.pmdb` instead of `headlamp.pmdb` as replacement on Linux which seems to be an incorrect behaviour.

**Comment by dnwillia-work** [comment]
> @dnwillia @seanpearsonuk Hi, would you be able to advise on fixing a build issue? The documentation build is failing on line 233 of `radiation_headlamp.py` where it calls `meshing.tui.mesh.check_mesh()` to check the mesh:
> 
> ```
> /home/ansys/actions-runner/_work/pyfluent/pyfluent/examples/00-fluent/radiation_headlamp.py failed leaving traceback:
> Traceback (most recent call last):
>   File "/home/ansys/actions-runner/_work/pyfluent/pyfluent/examples/00-fluent/radiation_headlamp.py", line 233, in <module>
>     meshing.tui.mesh.check_mesh()
>   File "/home/ansys/actions-runner/_work/_tool/Python/3.10.8/x64/lib/python3.10/site-packages/ansys/fluent/core/services/datamodel_tui.py", line 288, in __getattribute__
>     if PyMenu(self.service, path).get_child_names():
>   File "/home/ansys/actions-runner/_work/_tool/Python/3.10.8/x64/lib/python3.10/site-packages/ansys/fluent/core/services/datamodel_tui.py", line 166, in get_child_names
>     response = self._service.get_attribute_value(request)
>   File "/home/ansys/actions-runner/_work/_tool/Python/3.10.8/x64/lib/python3.10/site-packages/ansys/fluent/core/services/error_handler.py", line 15, in func
>     raise RuntimeError(ex.details()) from None
> RuntimeError: menu not found
> ```
> 
> I can't reproduce this on my local machine, which is running version `0.18.dev0` of `ansys-fluent-core`.

I think Mainak is right.  The CI/CD runs on Linux so you need to use the pmdb for the documentation build pipeline.

**Comment by ansjsia** [comment]
@mkundu1 @dnwillia-work It works correctly now, thanks!

**Comment by ansjsia** [comment]
@mkundu1 I'm getting this unit test fail on v23.2.0, but it's only occurring intermittently. It didn't seem to be happening yesterday. Do you have any insight on fixing this?

```
ERROR tests/test_flobject.py::test_find_children_from_settings_root_232 - RuntimeError: The connection to the Fluent server could not be established within the configurable 60 second time limit.
```

**Comment by raph-luc** [comment]
@ansjsia I can answer that, that is a random test failure being tracked here: https://github.com/ansys/pyfluent/issues/1747 

You can just re-run the failed tests when that happens

File: examples/00-fluent/radiation_headlamp.py [hunk 1]
```diff
@@ -0,0 +1,519 @@
...
+}
+rad_inp_bc.radiation_direction = [-0.848, 0, -0.53]
+
+###############################################################################
+# Set convergence criteria
+# ~~~~~~~~~~~~~~~~~~~~~~~~
+# Enable residual plots and set the convergence criteria to 'none'.
+
+solver.tui.solve.monitors.residual.plot("yes")
+solver.tui.solve.monitors.residual.criterion_type("3")
```
Comment (raph-luc) [inline examples/00-fluent/radiation_headlamp.py:455]: Do we have settings API commands to replace these?

File: examples/00-fluent/radiation_headlamp.py [hunk 2]
```diff
@@ -0,0 +1,519 @@
...
+# Check the mesh in meshing mode.
+
+meshing.tui.mesh.check_mesh()
+
+###############################################################################
+# Save mesh file
+# ~~~~~~~~~~~~~~
+# Save the mesh file (``headlamp.msh.h5``).
+
+meshing.tui.file.write_mesh("headlamp.msh.h5")
```
Comment (raph-luc) [inline examples/00-fluent/radiation_headlamp.py:243]: Starting with 2024 R1, this TUI call will be able to be replaced by `meshing.File.WriteMesh(FileName=)`, though more wordy and not using snake case

File: examples/00-fluent/radiation_headlamp.py [hunk 3]
```diff
@@ -0,0 +1,519 @@
...
+    },
+)
+volume_mesh_gen.Execute()
+
+###############################################################################
+# Check mesh in meshing mode
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~
+# Check the mesh in meshing mode.
+
+meshing.tui.mesh.check_mesh()
```
Comment (raph-luc) [inline examples/00-fluent/radiation_headlamp.py:236]: Doesn't look like there is a non-TUI equivalent to replace this one currently

File: examples/00-fluent/radiation_headlamp.py (hunk 2 above)
Comment (ansjsia) [reply examples/00-fluent/radiation_headlamp.py:243]: Would you recommend changing this line? I developed this on 2023 R2.

Comment (raph-luc) [reply examples/00-fluent/radiation_headlamp.py:243]: No, that was more of a note for myself and others. Currently our doc builds use Fluent 2023 R2 as well, so you used the right command there. I believe there is no non-TUI equivalent in this version.

File: examples/00-fluent/radiation_headlamp.py [hunk 4]
```diff
@@ -0,0 +1,519 @@
...
+solver = meshing.switch_to_solver()
+
+###############################################################################
+# Enable energy and viscosity models
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+# Set up the energy and viscosity models.
+
+solver.setup.models.energy.enabled = True
+
+solver.setup.models.viscous.model = "laminar"
```
Comment (raph-luc) [inline examples/00-fluent/radiation_headlamp.py:266]:
```suggestion
models = solver.setup.models
models.energy.enabled = True
models.viscous.model = "laminar"
```

File: examples/00-fluent/radiation_headlamp.py [hunk 5]
```diff
@@ -0,0 +1,519 @@
...
+###############################################################################
+# Set up radiation model
+# ~~~~~~~~~~~~~~~~~~~~~~
+# Set up the Monte Carlo radiation model. The number of histories is set to 10
+# million in order to reduce computation time, but this may need to be
+# increased to obtain accurate results.
+
+solver.setup.models.radiation.model = "monte-carlo"
+solver.setup.models.radiation.monte_carlo.number_of_histories = 1e7
+solver.setup.models.radiation.solve_frequency.iteration_interval = 20
```
Comment (raph-luc) [inline examples/00-fluent/radiation_headlamp.py:277]:
```suggestion
radiation = models.radiation
radiation.model = "monte-carlo"
radiation.monte_carlo.number_of_histories = 1e7
radiation.solve_frequency.iteration_interval = 20
```

**Review by raph-luc — COMMENTED** [review]
Minor suggestion

File: examples/00-fluent/radiation_headlamp.py (hunk 1 above)
Comment (raph-luc) [reply examples/00-fluent/radiation_headlamp.py:456]: I think something like this should work?
```
session.solution.monitor.residual.options.criterion_type = 3
session.solution.monitor.residual.plot() 
```

**Review by raph-luc — APPROVED** [review]
Looks great, thanks @ansjsia 

While building the docs for this example, I realized a few things seem to have changed about how all examples are displayed (or at least I didn't remember this behavior), created an issue tracker so we can look into this later: https://github.com/ansys/pyfluent/issues/2044

**Review by mkundu1 — APPROVED** [review]

**Review by prmukherj — APPROVED** [review]

File: examples/00-fluent/radiation_headlamp.py (hunk 1 above)
Comment (ansjsia) [reply examples/00-fluent/radiation_headlamp.py:456]: It doesn't seem to work on my setup, any luck with yours?
image
image

Comment (raph-luc) [reply examples/00-fluent/radiation_headlamp.py:456]: Nope, tried here and those that I identified above aren't direct equivalents. Seems like we do not have equivalents for those, thanks @ansjsia

**Comment by amscosta** [comment]
Hi,
When I try to run locally is crashing in the following line :
radiation = models.radiation

It says It can not find the .radiation model.
I am using 23.1
Any hint how to solve it ?

**Comment by ansjsia** [comment]
@amscosta You're likely getting that error because you're using 23.1, whereas this example was developed for 23.2-24.1. I'm not sure the radiation model is exposed for the 23.1 API.

**Comment by amscosta** [comment]
Thanks for the reply.
How can I verify that "exposition" ?
Would be great if each examples tells the version that is currently designed to work.

**Comment by seanpearsonuk** [comment]
> Thanks for the reply. How can I verify that "exposition" ? Would be great if each examples tells the version that is currently designed to work.

@amscosta we currently maintain the examples to be guaranteed to work with the latest release of Fluent.

**Pull request merged**